copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+        }
+
+        // Start sending all the individual updates
+        for (Map.Entry<ConnectorTaskId, Map<String, String>> taskConfigEntry : configs.entrySet()) {
+            Struct copycatConfig = new Struct(TASK_CONFIGURATION_V0);
+            copycatConfig.put("properties", taskConfigEntry.getValue());
+            byte[] serializedConfig = converter.fromCopycatData(topic, TASK_CONFIGURATION_V0, copycatConfig);
+            configLog.send(TASK_KEY(taskConfigEntry.getKey()), serializedConfig);
+        }
+
+        // Finally, send the root update, then wait until we read to the end of the log
+        try {
+            // Read to end to ensure all the task configs have been written
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+            Struct copycatConfig = new Struct(ROOT_V0);
+            copycatConfig.put("connectors", newRootConfig);
+            byte[] serializedConfig = converter.fromCopycatData(topic, ROOT_V0, copycatConfig);
+            configLog.send(ROOT_KEY, serializedConfig);
+
+            configLog.readToEnd().get(READ_TO_END_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            log.error("Failed to write root configuration to Kafka: ", e);
+            throw new CopycatException("Error writing root configuration to Kafka", e);
+        }
+    }
+
+    private KafkaBasedLog<String, byte[]> createKafkaBasedLog(String topic, Map<String, Object> producerProps,
+                                                              Map<String, Object> consumerProps, Callback<ConsumerRecord<String, byte[]>> consumedCallback) {
+        return new KafkaBasedLog<>(topic, producerProps, consumerProps, consumedCallback, new SystemTime());
+    }
+
+    private final Callback<ConsumerRecord<String, byte[]>> consumedCallback = new Callback<ConsumerRecord<String, byte[]>>() {
+        @Override
+        public void onCompletion(Throwable error, ConsumerRecord<String, byte[]> record) {
+            if (error != null) {
+                log.error("Unexpected in consumer callback for KafkaConfigStorage: ", error);
+                return;
+            }
+
+            final SchemaAndValue value;
+            try {
+                value = converter.toCopycatData(topic, record.value());
+            } catch (DataException e) {
+                log.error("Failed to convert config data to Copycat format: ", e);
+                return;
+            }
+
+            if (record.key().equals(ROOT_KEY)) {
+                List<ConnectorTaskId> updatedTasks;
+                synchronized (lock) {
+                    // Apply any outstanding deferred task updates and then update the root config. This *can* result in
+                    // an inconsistent output, which will have to be resolved by some external mechanism ensuring reconfiguration
```

Comment:
"external mechanism" means something outside this class. This basically means the herder. In theory, the only way this should happen is if you had a partial update to some tasks, the old config data for those tasks got compacted, and the writer of that partial update died before writing a new root update. When you start up a new node it could actually see a sequence of messages something like this:

```
connector foo, task 3 config: { ... }
connector foo, task 4 config: { ... }
root: { connector foo: 4 tasks }
connector foo, task 1 config: { ... }
connector foo, task 2 config: { ... }
```

This happens if we were trying to write a new set of task configs for `foo` but failed part way through. Now when we try to recover, only that first root update will be applied, but clearly we're missing some task configs. In that case, the herder should detect the inconsistent state and request that the necessary connectors be reconfigured and updated configs should be written.

The actual utility for determining if there is an inconsistency wasn't relevant to this patch yet since there's no code for reacting to the inconsistency yet. That'll come in the KAFKA-2471 patch and we can implement the consistency check in `ClusterConfigState` since that's the object that should actually be used by the herder.